### ISSUE TITLE
Warn on unused inline ignores

### ISSUE THREAD

**Issue by jpy-git** [sqlfluff/sqlfluff#2337]
This is a nice feature that `mypy` offers via
```cfg
[mypy]
warn_unused_ignores = True
```

The idea is that occasionally when using mypy you can run into a false positive (usually due to some incorrect/incomplete stubs in typeshed) and as a result you can add an inline ignore (`# type: ignore`) to ignore it.

However, over time these issues get fixed and your inline ignore can become redundant. With the flag above mypy will raise a lint for these redundant inline ignores, therefore allowing the user to clean them up from their code over time.

There are pretty obvious parallels here with how users might use our inline ignores to ignore a parse error or a misbehaving rule that we fix in a subsequent release but they never clean up the old inline ignores from their codebase.

In terms of implementing this I think there are a couple discussion points:
- When we parse the `NoQaDirective` it will be given some attribute like `exhausted=False` and when we then mask the lints we mark off the `NoQaDirective` as exhausted if it actually corresponds to a detected lint. Then flag the rest as unused and raise a lint for them.
- Like mypy this would not be enabled by default, it's more of an advanced feature for those who want stricter linting or are maybe doing a one off cleanup. It can be a config option + CLI flag like in mypy.
- OPTIONAL: We currently have a couple different syntaxes for inline ignoring that we could possibly simplify in favour of the simpler `mypy`/`flake8` style (i.e. just the `-- noqa: L006` style ones rather than pylint style directives) ([docs](https://docs.sqlfluff.com/en/stable/rules.html#inline-ignoring-errors)). This one's not essential, just my opinion based on how I've used the inline ignores and could be a good chance to remove them if they're not used since it would simplify this work.

**Comment by tunetheweb** [comment]
I like it!

One thing to consider is different versions of SQLFluff.

For example I use SQLFluff locally, but for some projects if use it via GitHub Super Linter in CI and that is usually a version or two behind (cause of our frequent release cadence and the fact they only update dependencies once a week or so).

So could get annoying if it warns on something I don’t want to remove yet.

Or maybe there could be a command line switch (I don’t want it in `.sqlfluffini` and that is shared between both) so can ignore those errors locally until CI catches up.

**Comment by jpy-git** [comment]
@tunetheweb Yeah it's easy enough to have it in either config or as a CLI flag without much extra work. Both use cases are valid.

mypy offers a CLI flag as well for the same reason.
![image](https://user-images.githubusercontent.com/80432516/149844681-203774c4-cdf3-477a-9a15-4f65d6c9dae8.png)
